Make games run on the proper IOS

Leseratte10 opened the pull request:
This commit makes sure that Brainslug reloads into the correct IOS that's requested by the game before booting. The old version of Brainslug didn't do that, which broke USB support in a couple games. 

Also, it fixes a bug where certain out-of-region discs wouldn't boot properly due to an invalid video mode.

Leseratte10 commented:
Looks like I added a bit too much code, seems as if the `apploader_skip_init_sequence` parameter is unused. Let me check that again ...

Chadderz121 commented:
Thanks for looking into this. A few high level concerns:
1. How do you know the IOS isn't in use when you reload? This is one of the big problems with switching IOSes.
2. Why do we need two separate (similar) apploaders? Is it that one is just an 'IOS number getter' and the other is actually loader? If so, could it be named more obviously perhaps?
3. Why did all the validation get removed in the apploader? Was it causing problems? The new code doesn't check return values for many methods.
4. Tabs and spaces seem inconsistent.

Leseratte10 commented:
1. What would it be in use for? You mean because there are the module and search background threads that run in the background that might be using IOS stuff? I did move the SD init code after the IOS reload because that was causing trouble on reload, but other than that it seems to be working. 
2. Yeah, that's what I added that second Apploader for. The IOSApploader just initializes the disc drive, reads as much info as needed to know what IOS the game needs, and then leaves the disc drive in a state where the actual Apploader code can continue reading the disc after the IOS Reload without causing the drive to spin down and then spin up again. You probably could have added that code to the main Apploader and use if clauses to skip certain parts if we're in the first (IOS only) stage but I thought it'd be cleaner this way. 
3. You mean the loops around the DI function calls until they're successful? Yeah, I removed them because I couldn't debug stuff when the program spams up all logs by re-trying over and over and over again. If a `DI_Init()` call fails once, and maybe twice, it is highly unlikely to succeed in the third or any other attempt. Maybe I'll add some more useful logic, like retrying once and then printing a useful error instead of just running into an endless loop that looks like a freeze to the end user. 
4. Dammit, looks like my editor messed that up, I'll go clean that up.

Leseratte10 commented:
1. I moved the Module and Search background call to after the IOS reload which should hopefully get rid of that potential issue.
2. Added comments to both apploader files to make that clearer.
3. I added your loop logic back.
4. Should now be back to all spaces.

Chadderz121 commented:
Thanks for the improvements.

Yes I can agree about the infinite retries unlikely to work, but I do believe some IOS calls benefit from at least one retry on failure (notably `IOS_ReloadIOS` can fail a few times iirc).
